### ISSUE TITLE
`using` is confusing and blocks a better alternative

### ISSUE THREAD

**Issue by JadenSimon** [microsoft/TypeScript#60172]
### Acknowledgement

- [X] I acknowledge that issues using this template may be closed without further explanation at the maintainer's discretion.

### Comment

`using`, on the surface, seems like a useful feature:
* Convenient and succinct syntax, `using r = new Resource()` is _clean_
* Standardization of disposal methods, no more `.exit`, `.close`, `.shutdown`, etc.
* Prevents common mistakes like failing to register for disposal immediately after initialization:
```ts
const r1 = new Resource()
const r2 = new Resource(r1) // If this fails, `r1` isn't cleaned up
```
vs.
```ts
using r1 = new Resource(), r2 = new Resource(r1) // All good!
```

These are are nice, _in theory_. Of course, `using` completely falls apart in practice:
* Adds friction by requiring a specific protocol (`Symbol.dispose` and `Symbol.asyncDispose`)
* Adds friction by required a binding, even if unused. It's not obvious how to dispose of a resource without `using` unless you know the internals of the feature.
* Because the protocol must be implemented on the target resource, consumers of libraries must do one of the following to use the feature:
    * Wait for the library to add support
    * Patch the library
    * Use a "convenience" API that is effectively equivalent to a block-scoped `defer` that other languages (Zig, Swift, and 
Go in a different form) have as direct syntax
* Limits imagination by being directly associated with the idea of "resources"
* Reduces readability by hiding behavior

Many of these points are discussed further in https://github.com/tc39/proposal-explicit-resource-management/issues/244#issue-2567542543:

In the above, I propose replacing `using` with a `defer` statement that executes code unconditionally on scope exit (multiple statements are last-in-first-out). Which looks like this:
```ts
const r = new Resource()
defer r.dispose()
```

While not being as succinct nor standardizing disposal methods, `defer` is far simpler compared to `using`. **I strongly believe that, if `using` becomes part of the spec, it will prevent `defer` from becoming apart of the language.** This concern is supported by the fact that the [Explicit Resource Management proposal](https://github.com/tc39/proposal-explicit-resource-management), which adds `using`, also adds an [inferior version of `defer`](https://github.com/tc39/proposal-explicit-resource-management?tab=readme-ov-file#interoperation-and-customization).

I have started working on a [proposal for `defer`](https://github.com/JadenSimon/proposal-defer) along with an implementation in my own transpiler to validate the concept.

## Example TypeScript Issues
All of these were effectively resolved as "expected behavior". This feature clearly does not align with user expectations. Also, it was difficult to find many issues related to `using`. It would be interesting to know how many people are using the feature.

* https://github.com/microsoft/TypeScript/issues/58544
`defer` significantly reduces the confusing here because it's not directly associated with variable bindings. It makes it clear that you are deferring the disposal of `Logger`. `using` misleads users, `defer` does not.

* https://github.com/microsoft/TypeScript/issues/55527
There would be no confusion with `defer` because it's not associated with variable bindings.

* https://github.com/microsoft/TypeScript/issues/55001
This issue is related to the nuances of async functions, `Promise.reject`, and how `using` is implemented in transpilers. `defer` does not directly address this problem, but I think it does a better job at addressing this comment:
> For someone unfamiliar with `using` it's not apparent that there's an "implicit finally section"

In summary, `defer` solves the same problems as `using` while being less confusing and without the baggage of protocols and resources. By supporting `using`, the language misses out on a superior alternative.

**Comment by fatcerberus** [comment]
The TypeScript repo isn't the place to discuss this. TypeScript simply follows the ES spec and doesn't add new runtime features to the language. `using` is an ECMAScript feature.

**Comment by JadenSimon** [comment]
> The TypeScript repo isn't the place to discuss this. TypeScript simply follows the ES spec and doesn't add new runtime features to the language. using is an ECMAScript feature.

The proposal is not apart of the spec yet. TypeScript could absolutely push back or raise concerns if it affects the usability of the tool.

**Comment by RyanCavanaugh** [comment]
The proposal repo is the proper place to raise these concerns

**Comment by JadenSimon** [comment]
@DanielRosenwasser 

I have serious concerns over the usability of TypeScript due to the support of `using` in the Explicit Resource Management proposal. I have already voiced my concerns in the proposal repository and have gotten no where. Is this the right place to discuss those concerns?

**Comment by RyanCavanaugh** [comment]
@JadenSimon the proposal author is meaningfully engaged with you in that. People not agreeing with you is not the same as them not listening to you.

**Comment by JadenSimon** [comment]
The proposal author largely attacked the concept of `defer` rather than addressing my usability concerns. I would hardly call that "meaningfully engaged". 

In any case, how this issue was handled says a lot. Thank you.
